File: SportTracker.App/Auth/AuthService.cs
using System.Net.Http.Json;
using System.Text.Json.Serialization;

namespace SportTracker.App.Auth;

/// <summary>
/// Appelle les endpoints <c>MapIdentityApi</c> (/login, /register) et gère la
/// persistance du token + la notification de l'état d'authentification.
/// </summary>
public class AuthService(
    HttpClient http,
    TokenStore store,
    CustomAuthenticationStateProvider authState)
{
    /// <summary>Connexion. Renvoie <c>null</c> si OK, sinon un message d'erreur.</summary>
    public async Task<string?> LoginAsync(string email, string password)
    {
        HttpResponseMessage response;
        try
        {
            response = await http.PostAsJsonAsync("login",
                new { email, password });
        }
        catch (HttpRequestException)
        {
            return "Impossible de joindre le serveur.";
        }

        if (!response.IsSuccessStatusCode)
            return "Email ou mot de passe incorrect.";

        var token = await response.Content.ReadFromJsonAsync<AccessTokenResponse>();
        if (token?.AccessToken is null)
            return "Réponse du serveur invalide.";

        await store.SetTokenAsync(token.AccessToken);
        authState.NotifyStateChanged();
        return null;
    }

    /// <summary>Inscription puis connexion automatique. Renvoie <c>null</c> si OK.</summary>
    public async Task<string?> RegisterAsync(string email, string password)
    {
        HttpResponseMessage response;
        try
        {
            response = await http.PostAsJsonAsync("register",
                new { email, password });
        }
        catch (HttpRequestException)
        {
            return "Impossible de joindre le serveur.";
        }

        if (!response.IsSuccessStatusCode)
        {
            var problem = await TryReadIdentityErrorsAsync(response);
            return problem ?? "Inscription impossible. Vérifie l'email et le mot de passe.";
        }

        // Identity renvoie 200 sans corps : on enchaîne sur un login.
        return await LoginAsync(email, password);
    }

    public async Task LogoutAsync()
    {
        await store.ClearAsync();
        authState.NotifyStateChanged();
    }

    private static async Task<string?> TryReadIdentityErrorsAsync(HttpResponseMessage response)
    {
        try
        {
            var problem = await response.Content.ReadFromJsonAsync<IdentityProblem>();
            if (problem?.Errors is { Count: > 0 } errors)
                return string.Join(" ", errors.Values.SelectMany(v => v));
        }
        catch
        {
            // corps non JSON / vide — on retombe sur le message générique
        }
        return null;
    }

    private sealed record AccessTokenResponse(
        [property: JsonPropertyName("accessToken")] string? AccessToken,
        [property: JsonPropertyName("expiresIn")] int ExpiresIn);

    private sealed record IdentityProblem(
        [property: JsonPropertyName("errors")] Dictionary<string, string[]>? Errors);
}

File: SportTracker.Web/src/api/authService.ts
import type { components } from './schema'
import { apiRequest, ApiError } from './client'
import { setDraftOwner, setToken } from './tokenStore'

type Credentials = components['schemas']['LoginRequest']
type RegisterRequest = components['schemas']['RegisterRequest']
type AccessTokenResponse = components['schemas']['AccessTokenResponse']
export type UserInfo = components['schemas']['InfoResponse']

function identityErrors(error: ApiError): string | null {
  const body = error.body as { errors?: Record<string, string[]> } | undefined
  const messages = Object.values(body?.errors ?? {}).flat()
  return messages.length ? messages.join(' ') : null
}

export async function login(email: string, password: string): Promise<string | null> {
  try {
    const response = await apiRequest<AccessTokenResponse>('/login', {
      method: 'POST', auth: false, body: { email, password } satisfies Credentials,
    })
    if (!response?.accessToken) return 'Réponse du serveur invalide.'
    setToken(response.accessToken, email)
    return null
  } catch (error) {
    if (error instanceof ApiError) return 'Email ou mot de passe incorrect.'
    return 'Impossible de joindre le serveur.'
  }
}

export async function register(email: string, password: string): Promise<string | null> {
  try {
    await apiRequest<void>('/register', {
      method: 'POST', auth: false, body: { email, password } satisfies RegisterRequest,
    })
  } catch (error) {
    if (error instanceof ApiError) return identityErrors(error) ?? "Inscription impossible. Vérifie l'email et le mot de passe."
    return 'Impossible de joindre le serveur.'
  }
  return login(email, password)
}

export async function getUserInfo(): Promise<UserInfo> {
  const info = await apiRequest<UserInfo>('/manage/info')
  setDraftOwner(info.email)
  return info
}
